Description: socket timeout support float.
 float value is supported by php5
 (http://php.net/manual/en/function.stream-socket-client.php)
Author: Tang Hongbo <thbourlove@gmail.com>
Reviewed-by: Giuseppe Lavagetto <glavagetto@wikimedia.org>
Last-Update: 2016-01-29
Origin: https://github.com/facebook/hhvm/pull/5222

diff --git a/hphp/runtime/ext/sockets/ext_sockets.cpp b/hphp/runtime/ext/sockets/ext_sockets.cpp
index 56da92b..7f18fcd 100644
--- a/hphp/runtime/ext/sockets/ext_sockets.cpp
+++ b/hphp/runtime/ext/sockets/ext_sockets.cpp
@@ -351,7 +351,7 @@ static SmartPtr<Socket> create_new_socket(
 }
 
 static int connect_with_timeout(int fd, struct sockaddr *sa_ptr,
-                                size_t sa_size, int timeout,
+                                size_t sa_size, double timeout,
                                 const HostURL &hosturl,
                                 std::string &errstr, int &errnum) {
   if (timeout <= 0) {
@@ -402,7 +402,7 @@ static int connect_with_timeout(int fd, struct sockaddr *sa_ptr,
   return retval;
 }
 
-static Variant new_socket_connect(const HostURL &hosturl, int timeout,
+static Variant new_socket_connect(const HostURL &hosturl, double timeout,
                                   const SmartPtr<StreamContext>& streamctx,
                                   Variant &errnum, Variant &errstr) {
   int domain = AF_UNSPEC;
diff --git a/hphp/test/slow/ext_socket/client_float_timeout_open_port.php b/hphp/test/slow/ext_socket/client_float_timeout_open_port.php
new file mode 100644
index 0000000..7650106
--- /dev/null
+++ b/hphp/test/slow/ext_socket/client_float_timeout_open_port.php
@@ -0,0 +1,22 @@
+<?php
+function random_free_port() {
+  for ($i = 0; $i < 100; $i++) {
+    $port = rand(50000, 65000);
+    if ($socket = @socket_create_listen($port)) {
+      socket_close($socket);
+      return $port;
+    }
+  }
+  return 0;
+}
+
+$port = random_free_port();
+$time = microtime(true);
+@stream_socket_client("tcp://172.0.0.1:$port", $errno, $errstr, 0.001);
+$elapsed = microtime(true) - $time;
+echo $errstr, "\n";
+if ($elapsed < 1) {
+  print "SUCCESS";
+} else {
+  print "FAILURE";
+}
diff --git a/hphp/test/slow/ext_socket/client_float_timeout_open_port.php.expect b/hphp/test/slow/ext_socket/client_float_timeout_open_port.php.expect
new file mode 100644
index 0000000..176553d
--- /dev/null
+++ b/hphp/test/slow/ext_socket/client_float_timeout_open_port.php.expect
@@ -0,0 +1,2 @@
+Connection timed out
+SUCCESS
